tfe/resource_tfe_variable_set.go

```diff
@@ -0,0 +1,181 @@
+package tfe
+
+import (
+	"fmt"
+	"log"
+	"regexp"
+
+	tfe "github.com/hashicorp/go-tfe"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+var variableSetIdRegexp = regexp.MustCompile("varset-[a-zA-Z0-9]{16}$")
+
+func resourceTFEVariableSet() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceTFEVariableSetCreate,
+		Read:   resourceTFEVariableSetRead,
+		Update: resourceTFEVariableSetUpdate,
+		Delete: resourceTFEVariableSetDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"global": {
+				Type:          schema.TypeBool,
+				Optional:      true,
+				Default:       false,
+				ConflictsWith: []string{"workspace_ids"},
+			},
+
+			"organization": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"workspace_ids": {
```

Comment:
Referencing my post in this thread: https://github.com/hashicorp/terraform-provider-tfe/issues/391#issuecomment-1036929310

By coupling the variable set creation logic, with the assignment logic, this resource becomes much harder to consume.

This means that you have to create the variable set and assign it to child workspaces, all in the same workspace.

Is it possible to further split tfe_variable_set resource further to decouple the logic?

Reply:
This makes sense to me, I definitely think the `workspace_ids` attribute being both a computed property and an argument can make it harder to reason about your configuration, especially if that variable set is attached to a workspace outside your configuration. Your config might look like:

```hcl
resource "tfe_variable_set" "foo" {
  # other attributes here
  workspace_ids = ["ws-ID1"]
}
```
and your state file: 
```json
"workspace_ids": ["ws-ID1", "ws-ID2", "ws-ID3"] 
```

_(which I think it's weird 'cause it creates a disjunction between intent and reality)_ 

Expressing relationships as resources has been done before, the examples I can think of are: `tfe_organization_module_sharing` and `tfe_organization_membership`. Doing so for variable sets, we can at least be explicit about what VS->Wkspace relationships are managed by Terraform, while keeping `workspace_ids` as the source of truth for _all_ relationships (including those not managed by Terraform).

Reply:
Appreciate this PR is closed. Now that you have explained it, I can understand why that could be difficult to work with.

The Vault Provider deals with this with the use of an `explicit = true` and `external_policies = true` flag, as seen in these two resources:
[identity_group_policies](https://registry.terraform.io/providers/hashicorp/vault/latest/docs/resources/identity_group_policies)
[identity_group_policies](https://registry.terraform.io/providers/hashicorp/vault/latest/docs/resources/identity_group)

This tells Terraform to ignore any items that weren't previously managed under state to avoid the disfunction you spoke of earlier. I haven't looked closely enough at the provider code to understand how that works under the hood. 

You mentioned you want to decouple the conversation. Please advise where you would like to track that going forth if looking for further input :)